x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java

```diff
@@ -56,4 +104,83 @@ public Query wrapFunctionQuery(ScalarFunction sf, Expression field, Supplier<Que
+                bc.right().sourceLocation().getLineNumber(),
+                bc.right().sourceLocation().getColumnNumber(),
+                Expressions.name(bc.right()),
+                bc.symbol()
+            );
+        }
+
+        static Query translate(InsensitiveEquals bc) {
+            TypedAttribute attribute = checkIsPushableAttribute(bc.left());
+            Source source = bc.source();
+            Object value = ExpressionTranslators.valueOf(bc.right());
+            String format = null;
+            boolean isDateLiteralComparison = false;
+
+            // for a date constant comparison, we need to use a format for the date, to make sure that the format is the same
+            // no matter the timezone provided by the user
+            if (value instanceof ZonedDateTime || value instanceof OffsetTime) {
+                DateFormatter formatter;
+                if (value instanceof ZonedDateTime) {
+                    formatter = DateFormatter.forPattern(ExpressionTranslators.DATE_FORMAT);
+                    // RangeQueryBuilder accepts an Object as its parameter, but it will call .toString() on the ZonedDateTime instance
+                    // which can have a slightly different format depending on the ZoneId used to create the ZonedDateTime
+                    // Since RangeQueryBuilder can handle date as String as well, we'll format it as String and provide the format as well.
+                    value = formatter.format((ZonedDateTime) value);
+                } else {
+                    formatter = DateFormatter.forPattern(ExpressionTranslators.TIME_FORMAT);
+                    value = formatter.format((OffsetTime) value);
+                }
+                format = formatter.pattern();
+                isDateLiteralComparison = true;
+            } else if (attribute.dataType() == IP && value instanceof BytesRef bytesRef) {
+                value = DocValueFormat.IP.format(bytesRef);
+            } else if (attribute.dataType() == VERSION) {
+                // VersionStringFieldMapper#indexedValueForSearch() only accepts as input String or BytesRef with the String (i.e. not
+                // encoded) representation of the version as it'll do the encoding itself.
+                if (value instanceof BytesRef bytesRef) {
+                    value = new Version(bytesRef).toString();
+                } else if (value instanceof Version version) {
+                    value = version.toString();
+                }
+            } else if (attribute.dataType() == UNSIGNED_LONG && value instanceof Long ul) {
+                value = unsignedLongAsNumber(ul);
+            }
+
+            ZoneId zoneId = null;
+            if (DataTypes.isDateTime(attribute.dataType())) {
+                zoneId = bc.zoneId();
+            }
+
+            String name = pushableAttributeName(attribute);
+
+            Query query;
+            if (isDateLiteralComparison) {
```

Comment:
The operator should fallback to equals for other type so the other types should not make a difference.
IPs fields would be handled separately even when dealing with a string value:

`ip_field =~ "2001:0:5ef5..."` should be the same as `ip_field == "2001:0:5ef5..."`

The  IPv6 comparisons itself should case-insensitive but at this point we're discussing about IP fields not =~

Reply:
Yeah - ipv6 is case insensitive all the time because we're just parsing to the ip. It's insensitive of normalization too.

I dunno, I kind of feel like `=~` just shouldn't work on non-string fields.

Reply:
>  I kind of feel like =~ just shouldn't work on non-string fields.
Makes it easier to reason about so let's do that and throw a validation error in other cases.